Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignature.swift

```diff
@@ -16,130 +16,313 @@ import SymbolKit
+        while !markedAngleBracketPairs.isEmpty {
+            let (open, close) = markedAngleBracketPairs.removeLast()
+            assert(self[open] == openAngle, "Start marker at \(open) is '\(String(cString: [self[open], 0]))' instead of '>' in \(String(cString: self + [0]))")
+            assert(self[close] == closeAngle, "End marker at \(close) is '\(String(cString: [self[close], 0]))' instead of '>' in \(String(cString: self + [0]))")
             
-            if scanner.hasPrefix("Array<") {
-                _ = scanner.take("Array".count)
-                guard let elementType = scanner.scanGenericSingle() else {
-                    // The type is unexpected. Return the original value.
-                    return String(self)
+            // The caller accumulated a single character for each marker that indicated the type of syntactic sugar to apply.
+            let marker = open - 1
+            switch self[marker] {
+                
+            case uppercaseA: // Array
+                // Apply Swift array syntactic sugar; transforming "A<Element>" into "[Element]" (where "Array" was already abbreviated to "A").
+                self[close] = closeSquare
+                self.replaceSubrange(marker ... open /* "A<" */, with: [openSquare])
+                
+                // Update later marked locations since the syntactic sugar shortened the string.
+                for index in markedAngleBracketPairs.indices where open < markedAngleBracketPairs[index].close {
+                    markedAngleBracketPairs[index].close -= 1 // "A<" is replaced by "["
+                    // The `open` location doesn't need to be updated because the pairs are iterated over in reverse order
                 }
-                result.append("[\(elementType.withSwiftSyntacticSugar())]")
-                assert(scanner.peek() == ">")
-                _ = scanner.take()
-            }
-            else if scanner.hasPrefix("Optional<") {
-                _ = scanner.take("Optional".count)
-                guard let wrappedType = scanner.scanGenericSingle() else {
-                    // The type is unexpected. Return the original value.
-                    return String(self)
+                
+            case uppercaseO: // Optional
+                // Apply Swift optional syntactic sugar; transforming "O<Wrapped>" into "Wrapped?" (where "Optional" was already abbreviated to "O").
+                self[close] = question
+                self.removeSubrange(marker ... open /* "O<" */)
+                
+                // Update later marked locations since the syntactic sugar shortened the string.
+                for index in markedAngleBracketPairs.indices where open < markedAngleBracketPairs[index].close {
+                    markedAngleBracketPairs[index].close -= 2 // "O<" is removed
+                    // The `open` location doesn't need to be updated because the pairs are iterated over in reverse order
```

Comment:
The code across both these cases is very similar (and also used for the `uppercaseD` case), it would be beneficial to extract it into a separate (private) function.

Reply:
In my opinion the common pieces are too small to merit separate functions. 

If you look closely at the code there are subtle differences between the cases. For example, comparing the Array case with the Optional case;
- The Array case replaces the closing bracket with `]` but the Optional case replaces it with `?`
- The Array case replaces the opening bracket with `[` but the Optional case removes it
- The Array case offsets the later markers by `-1` but the Optional case offset  them by `-2`

In other words, 3 out of 4 lines (I'm not counting `}`) have differences between those cases.

The Dictionary case is close to the array case but it also needs to find and replace `,` with `:`.

A common function that deals with all cases would in my opinion have too many parameters for the little reuse. 

Also, because the shared code couldn't be specific to any of the cases it would miss out on tiny inline comments like this that provide context to what these ranges represent. 
```swift
self.replaceSubrange(marker ... open /* "A<" */, with: [openSquare])
```

This in my opinion would make the code harder to read than 3 separate cases with specific comments in each case.